### PR TITLE
Handle big size image

### PR DESCRIPTION
**Pull request by JerryShih**
@nical @glennw @kvark @sotaroikeda 
r?

This patch try to handle the #1405.
If we hit the maximum texture size, turn to use a 1x1 rgba(255,255,255,255) dummy texture.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/1564)

### CODE REVIEW

**Comment by glennw** [comment]
I'm working on a large refactoring of the texture cache at the moment - if possible, I'd like to hold off on merging this until that's done. Is that going to be OK? (I would expect it to be done early next week). Handling this case should be *much* simpler with the changes I'm working on, too.

**Comment by JerryShih** [comment]
@glennw 
That's fine. If the refactor is done, please show the note in this thread.

**Comment by glennw** [comment]
@JerryShih The texture cache refactor has landed now. Instead of trying to use a dummy texture for this case, could we just not add the item to any batches when we find an invalid texture?

**Comment by JerryShih** [comment]
> @JerryShih The texture cache refactor has landed now. Instead of trying to use a dummy texture for this case, could we just not add the item to any batches when we find an invalid texture?

@glennw 
That could be another approach. I think I will still use a special id for the big size allocation(or failed allocation). Then, we could skip the item with that id in batch list or draw with a special color texture for debugging(with a 1x1 black dummy texture).

**Comment by glennw** [comment]
I think the simplest approach will be to early exit from ```request_image``` in resource cache, before even getting to texture cache. Then, ```get_cached_image``` could just return ```SourceTexture::Invalid```. But let's see how this looks after rebasing to the new texture cache and decide from there :)

**Comment by JerryShih** [comment]
@glennw r?

And I need the help for the wrench reftest. I got the different result from the standalone yaml replay and the reftest result.

standalone:
I could see a black Rect. That's what I want.
`cargo run show reftests/image/very-big-tile-size.yaml`

reftest:
I got a gray Rect.
`./headless.py reftest reftests/image/very-big-tile-size.yaml`

I don't know the difference between them.

**Comment by JerryShih** [comment]
@glennw r? again.
rebase.
I still hit the reftest problem with wrench. Maybe you have some hints for this problem.

**Comment by glennw** [comment]
@JerryShih I'm not really sure why it would be different between standalone and reftest. Perhaps the wrong texture is getting bound or something like that?

I wrote up a prototype of what I was discussing above here https://github.com/glennw/webrender/commit/9ca1214d73e658499e599e73ec200668ad6d17fc

I think if we use a solution like this, where it doesn't involve the texture cache at all, the code is quite a bit simpler. And when we want to display the bad texture, we could do that as a follow up in the renderer (handle the case where a texture resolves to SourceTexture::Invalid during batching). What do you think?

**Comment by JerryShih** [comment]
> @JerryShih I'm not really sure why it would be different between standalone and reftest. Perhaps the wrong texture is getting bound or something like that?
> 
> I wrote up a prototype of what I was discussing above here glennw/webrender@9ca1214
> 
> I think if we use a solution like this, where it doesn't involve the texture cache at all, the code is quite a bit simpler. And when we want to display the bad texture, we could do that as a follow up in the renderer (handle the case where a texture resolves to SourceTexture::Invalid during batching). What do you think?

Your prototype is much simpler than me. So, yours is a better solution. I add a comment in glennw/webrender@9ca1214. I will create another follow up if we need to show the skipped image.

**Comment by glennw** [comment]
@JerryShih Sounds good, thanks! Do you want me to follow up your comments and open a PR, or should I leave that for you?

**Comment by JerryShih** [comment]
@glennw 
I would like to do that!

**Comment by JerryShih** [comment]
@glennw has a better solution. Close this pr.

**Comment by JerryShih** [comment]
@glennw @kvark r? again.
This version just skip the image request and return a invalid CacheItem for the big size image request.

File: webrender/src/resource_cache.rs [hunk 1]
```diff
@@ -450,6 +460,23 @@ impl ResourceCache {
                     return;
                 }
 
+                if let Some(tile_size) = template.tiling {
+                    // If the tiling size is too big for hardware texture size then
+                    // just early out and drop the image.
+                    if tile_size as u32 > self.texture_cache.max_texture_size() {
+                        warn!("Dropping image, tile size:{} is too big for hardware!", tile_size);
+                        return;
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: I'm a bit uncomfortable with these `return` statements as they are different from other return code paths from this method in that they are failures, while others are successes.
How about we return true/false (at least), and then also modify `prepare_prim_for_render` to return `Option<Metadata>` so that a primitive that failed to prepare can be discarded in time?

File: webrender/src/resource_cache.rs [hunk 2]
```diff
@@ -546,13 +573,14 @@ impl ResourceCache {
...
 
         for (loop_index, key) in glyph_keys.iter().enumerate() {
-            let glyph = glyph_key_cache.get(key);
-            let cache_item = glyph.as_ref().map(|info| self.texture_cache.get(&info.texture_cache_handle));
-            if let Some(cache_item) = cache_item {
-                f(loop_index, &cache_item.uv_rect_handle);
-                debug_assert!(texture_id == None ||
-                              texture_id == Some(cache_item.texture_id));
-                texture_id = Some(cache_item.texture_id);
+            if let Some(glyph) = glyph_key_cache.get(key) {
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: We need to handle the case of failure here. Does it make sense to return as usual if some of the glyphs failed to be obtained?

File: webrender/src/resource_cache.rs [hunk 3]
```diff
@@ -588,7 +616,15 @@ impl ResourceCache {
...
         };
         let image_info = &self.cached_images.get(&key);
-        self.texture_cache.get(&image_info.texture_cache_handle)
+
+        // If an image is not in cached_images, just return an invalid CacheItem
+        // for that request. That invalid item will be skipped during the rendering.
+        //
+        // TODO(Jerry): add a debug option to fill the corresponding area for
+        // that invalid CacheItem.
+        image_info.map_or(CacheItem::invalid(), |image_info| {
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: we call `get_glyphs` when populating batches. Wouldn't it be simpler to just return an error here, so that a corresponding instance can be discarded?

File: webrender/src/resource_cache.rs (hunk 1 above)
Comment (JerryShih) [reply webrender/src/resource_cache.rs]: @kvark 
I'm not familiar with the frame_builder::build process. The prepare_prim_for_render() is called by frame_builder::handle_primitive_run()[1]. And the handle_primitive_run() looks like to handle the mask and clip related thing. Are you thinking about skipping the render-task creation in handle_primitive_run() for the case of None Option<Metadata> return value? 

[1]
https://dxr.mozilla.org/mozilla-central/rev/c959327c6b75cd4930a6ea087583c38b805e7524/gfx/webrender/src/frame_builder.rs#1996

Comment (kvark) [reply webrender/src/resource_cache.rs]: Yes

**Comment by glennw** [comment]
@JerryShih @kvark What's the status of this?

**Comment by JerryShih** [comment]
@glennw I'm check an animation artifact problem with WR. I will go back to finish this soon.

File: webrender/src/tiling.rs [hunk 4]
```diff
@@ -813,6 +817,11 @@ impl ClipBatcher {
                     ClipSource::Image(ref mask) => {
                         let cache_item =
                             resource_cache.get_cached_image(mask.image, ImageRendering::Auto, None);
+                        // Skip the invalid image mask.
+                        if cache_item.texture_id == SourceTexture::Invalid {
```
Comment (JerryShih) [inline webrender/src/tiling.rs]: This image mask will be rendered at [1]. So, I think we should skip this image if the texture id is not valid.

[1]
https://github.com/servo/webrender/blob/e11297a240b29bd57154d37ecb402d0b9aa299e6/webrender/src/renderer.rs#L2912

File: webrender/src/resource_cache.rs (hunk 3 above)
Comment (JerryShih) [reply webrender/src/resource_cache.rs]: We call get_glyphs at [1]. If there is any error inside get_glyphs()[2], the lambda expression at [3] will not be called(which will push the instance to draw list).

[1]
https://github.com/servo/webrender/blob/e11297a240b29bd57154d37ecb402d0b9aa299e6/webrender/src/tiling.rs#L513
[2]
https://github.com/servo/webrender/pull/1564/files#diff-77cbdf7ba9ebae81feb38a64c21b8454R630
[3]
https://github.com/servo/webrender/blob/e11297a240b29bd57154d37ecb402d0b9aa299e6/webrender/src/tiling.rs#L518

File: webrender/src/resource_cache.rs (hunk 1 above)
Comment (JerryShih) [reply webrender/src/resource_cache.rs]: The prepare_prim_for_render() is changed. The "Metadata" can't be remove now.

From the first comment:

> I'm a bit uncomfortable with these return statements as they are different from other return code paths from this method in that they are failures, while others are successes.

We still have a case with a failure at [1] in request_image(). So, the new return statements are not just the new failed case. If they are still not looked good, I need to check the  prepare_prim_for_render() for the "Metadata" removing.

[1]
https://github.com/servo/webrender/blob/e11297a240b29bd57154d37ecb402d0b9aa299e6/webrender/src/resource_cache.rs#L553

**Comment by JerryShih** [comment]
@glennw @kvark r?

File: webrender/src/resource_cache.rs [hunk 5]
```diff
@@ -127,10 +136,11 @@ where
         }
     }
 
-    fn get(&self, key: &K) -> &V {
-        self.resources
-            .get(key)
-            .expect("Didn't find a cached resource with that ID!")
+    /// In some cases, we can't handle the resource request(e.g. the user
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: I think we should still differentiate between keys that didn't end up in the map for some other logical reasons, and keys that were just failing to be added. Perhaps, the resources map should contain `Result<V>` instead of `V`, so that we can handle those failures in a more explicit way?

File: webrender/src/prim_store.rs [hunk 6]
```diff
@@ -1057,7 +1057,6 @@ impl PrimitiveStore {
         }
     }
 
-    /// Returns true if the bounding box needs to be updated.
```
Comment (JerryShih) [inline webrender/src/prim_store.rs]: This comment is not correct now.

File: webrender/src/resource_cache.rs [hunk 7]
```diff
@@ -113,8 +122,14 @@ struct CachedImageInfo {
...
 
+pub enum ResourceClassCacheError {
+    OverLimitSize,
+}
+
+pub type ResourceCacheCacheResult<V> = Result<V, ResourceClassCacheError>;
+
 pub struct ResourceClassCache<K, V> {
-    resources: FastHashMap<K, V>,
+    resources: FastHashMap<K, ResourceCacheCacheResult<V>>,
```
Comment (JerryShih) [inline webrender/src/resource_cache.rs]: Use Result<V> instead of V.

File: webrender/src/resource_cache.rs [hunk 8]
```diff
@@ -483,25 +498,46 @@ impl ResourceCache {
                     return;
                 }
 
+                if let Some(tile_size) = template.tiling {
+                    // If the tiling size is too big for hardware texture size then
+                    // just early out and drop the image.
+                    if tile_size as u32 > self.texture_cache.max_texture_size() {
+                        warn!("Dropping image, tile size:{} is too big for hardware!", tile_size);
+                        self.cached_images.insert(request, Err(ResourceClassCacheError::OverLimitSize));
```
Comment (JerryShih) [inline webrender/src/resource_cache.rs]: If we hit the size limitation, just insert an Err() entry.

**Comment by JerryShih** [comment]
@kvark r?
Try to use Result<T> in ResourceClassCache.

**Comment by JerryShih** [comment]
I haven't handled the Err() case for each get/get_mut() call, but I will just add some logs for these paths if we need to have these logs.

File: webrender/src/resource_cache.rs [hunk 9]
```diff
@@ -113,8 +122,14 @@ struct CachedImageInfo {
     epoch: Epoch,
 }
 
+pub enum ResourceClassCacheError {
+    OverLimitSize,
+}
+
+pub type ResourceCacheCacheResult<V> = Result<V, ResourceClassCacheError>;
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: perhaps you meant `ResourceClassCacheResult`?

File: webrender/src/resource_cache.rs [hunk 10]
```diff
@@ -159,7 +174,7 @@ where
             .cloned()
             .collect::<Vec<_>>();
         for key in resources_to_destroy {
-            self.resources.remove(&key).unwrap();
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: why is this removed? I assume we get a warning about unused `Result` or `Option`

File: webrender/src/resource_cache.rs [hunk 11]
```diff
@@ -483,25 +498,46 @@ impl ResourceCache {
...
+                    // If the tiling size is too big for hardware texture size then
+                    // just early out and drop the image.
+                    if tile_size as u32 > self.texture_cache.max_texture_size() {
+                        warn!("Dropping image, tile size:{} is too big for hardware!", tile_size);
+                        self.cached_images.insert(request, Err(ResourceClassCacheError::OverLimitSize));
+                        return;
+                    }
+                } else {
+                    // The image is too big for hardware texture size.
+                    if template.descriptor.width > self.texture_cache.max_texture_size() ||
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: should probably merge this block with the one above, using `let side_size = image.tiling.unwrap_or(cmp::max(template.descriptor.width, template.descriptor.height));`

File: webrender/src/resource_cache.rs [hunk 12]
```diff
@@ -483,25 +498,46 @@ impl ResourceCache {
...
+                    }
+                }
+
                 // If this image exists in the texture cache, *and* the epoch
                 // in the cache matches that of the template, then it is
                 // valid to use as-is.
                 let (entry, needs_update) = match self.cached_images.entry(request) {
                     Occupied(entry) => {
-                        let needs_update = entry.get().epoch != template.epoch;
+                        let needs_update = entry.get().as_ref().ok().unwrap().epoch != template.epoch;
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: no need for `ok()` here

File: webrender/src/resource_cache.rs [hunk 13]
```diff
@@ -483,25 +498,46 @@ impl ResourceCache {
...
+                                texture_cache_handle: TextureCacheHandle::new(),
+                            }
+                        )),
                         true,
                     ),
                 };
 
                 let needs_upload = self.texture_cache
-                    .request(&mut entry.texture_cache_handle, gpu_cache);
+                    .request(&mut entry.as_mut().ok().unwrap().texture_cache_handle, gpu_cache);
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: same here - no need for `ok()`

File: webrender/src/resource_cache.rs [hunk 14]
```diff
@@ -656,7 +692,15 @@ impl ResourceCache {
             tile,
         };
         let image_info = &self.cached_images.get(&key);
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: would probably be cleaner with a simple `match` statement

File: webrender/src/resource_cache.rs [hunk 15]
```diff
@@ -816,7 +860,7 @@ impl ResourceCache {
                 image_template.descriptor.clone()
             };
 
-            let entry = self.cached_images.get_mut(&request).unwrap();
+            let entry = self.cached_images.get_mut(&request).as_mut().ok().unwrap();
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: no need for `ok()`

File: webrender/src/tiling.rs [hunk 16]
```diff
@@ -516,6 +517,8 @@ impl AlphaRenderItem {
                             glyph_fetch_buffer,
                             gpu_cache,
                             |texture_id, glyphs| {
+                                debug_assert!(texture_id != SourceTexture::Invalid);
```
Comment (kvark) [inline webrender/src/tiling.rs]: nit: `debug_assert_ne!()`

File: webrender/src/resource_cache.rs [hunk 17]
```diff
@@ -656,7 +692,15 @@ impl ResourceCache {
...
         };
         let image_info = &self.cached_images.get(&key);
-        self.texture_cache.get(&image_info.texture_cache_handle)
+
+        // If an image is not in cached_images, just return an invalid CacheItem
+        // for that request. That invalid item will be skipped during the rendering.
+        //
+        // TODO(Jerry): add a debug option to fill the corresponding area for
+        // that invalid CacheItem.
+        image_info.as_ref().ok().map_or(CacheItem::invalid(), |image_info| {
```
Comment (kvark) [inline webrender/src/resource_cache.rs]: let's just return `Result<CacheItem, _>` from this function and match it on the other side, there is no need for nullable `CacheItem` semantic (that `invalid()` does here)

File: webrender/src/resource_cache.rs [hunk 18]
```diff
@@ -174,7 +167,7 @@ where
             .cloned()
             .collect::<Vec<_>>();
         for key in resources_to_destroy {
-            self.resources.remove(&key);
+            self.resources.remove(&key).unwrap();
```
Comment (JerryShih) [inline webrender/src/resource_cache.rs]: I revert the change, but the CI test is failed.

error: unused `std::result::Result` which must be used
   --> src/resource_cache.rs:170:13
    |
170 |             self.resources.remove(&key).unwrap();
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D unused-must-use` implied by `-D warnings`

**Comment by JerryShih** [comment]
@kvark r? for https://github.com/servo/webrender/pull/1564/commits/d68cf51bd7d9065e24688d15b772ea5fd58ebeab

**Comment by kvark** [comment]
Thank you for addressing my comments!
 >self.resources.remove(&key).unwrap();

Let's do `let _ = self.resources.remove(&key).unwrap();`

Please squash the commits together, and we can merge once CI is happy.

**Comment by JerryShih** [comment]
@kvark 
Update for CI "self.resources.remove(&key).unwrap();" failed, and squash the patch set.

**Comment by kvark** [comment]
@JerryShih awesome, :shipit: ! 👍 
@bors-servo r+
